src/plugins/embeddable/public/lib/embeddables/attribute_service.ts

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { SavedObjectsClientContract } from '../../../../../core/public';
+import {
+  SavedObjectEmbeddableInput,
+  isSavedObjectEmbeddableInput,
+  EmbeddableInput,
+  IEmbeddable,
+} from '.';
+import { SimpleSavedObject } from '../../../../../core/public';
+
+export class AttributeService<
+  SavedObjectAttributes,
+  ValType extends EmbeddableInput & { attributes: SavedObjectAttributes },
+  RefType extends SavedObjectEmbeddableInput
+> {
+  constructor(private type: string, private savedObjectsClient: SavedObjectsClientContract) {}
+
+  public async unwrapAttributes(input: RefType | ValType): Promise<SavedObjectAttributes> {
```

Comment:
We assume that saved object's attributes are matching embeddable input, correct? 

But this isn't the case, for example, for `dashboardContainer`, right? 

I wonder if we should keep this in mind and allow to provide custom mappers? Or are we consider `dashboardContainer` a `legacy` example?

Reply:
We aren't assuming that the saved object's attributes match the attributes key in embeddable input, because `attribute_service` actually  _requires_ it (which comes with its own problems). 

When creating an instance of `attribute_service`, you need to provide a type argument for the saved object attributes `AttributeType` but also, the value type argument must extend `{ attributes: AttributeType }`. This means that any embeddable that wants to make use of this service must follow an exact convention. 

I certainly wouldn't say that this convention would mark any other types of embeddable as `legacy`. The idea is that it's more of an optional service (helper maybe?) to make it easier to adopt the 'by value OR by reference' paradigm.

The custom mappers you suggest might make this a little more flexible. For instance, a `valueAttributesMapper` function param which takes the 'by value' input and returns the attributes, but it seems like that would require method level type parameters instead of the current class level ones.

Do you have any thoughts on  if it would be best to generalize this service, or leave it as an optional helper with very stringent requirements?

@stacey-gammon

Reply:
@ppisljar, can you look into this and provide your thoughts?

Reply:
i am wondering if there is any existing embeddable this can be used for ? or do we know for sure of future embddable that would need this ? it feels we don't have the scenario where embeddable input would equal saved object attributes, also loading of saved objects in most cases happen in special way (is not general).

if that is the case then i would vote for not having this service at all and leave it up to the embeddable implementation to do this (if that is anyway what they would need to do, at least for current embeddables)